### PR TITLE
[Merged by Bors] - Optimise attestation selection proof signing

### LINKED ISSUE #3963: Produced Orphaned block
We have produced a forked/orfan block:
https://beaconcha.in/slot/5770016
Validator: 338395
Slot: 5770016
Epoch: 180313

We have a 3-layer configuration with the validator client, the Beacon node and the execution client running separately.
We are using Mev-boost.
The setup is working well before and after the orfand block produced. So the problem is related to something "casual"

```

node07         | Feb 10 16:30:36.495 INFO Successfully published block            slot: 5768551, graffiti: Some("Block by Stakely Lido07"), attestations: 97, deposits: 0, block_type: Blinded, service: block
node07         | Feb 10 19:14:48.068 INFO Successfully published block            slot: 5769372, graffiti: Some("Block by Stakely Lido07"), attestations: 128, deposits: 0, block_type: Blinded, service: block
node07         | Feb 10 19:20:35.990 INFO Successfully published block            slot: 5769401, graffiti: Some("Block by Stakely Lido07"), attestations: 77, deposits: 0, block_type: Blinded, service: block
node07         | Feb 10 21:23:41.061 INFO Successfully published block            slot: 5770016, graffiti: Some("Block by Stakely Lido07"), attestations: 74, deposits: 0, block_type: Blinded, service: block
node07         | Feb 10 22:49:00.007 INFO Successfully published block            slot: 5770443, graffiti: Some("Block by Stakely Lido07"), attestations: 105, deposits: 0, block_type: Blinded, service: block
node07         | Feb 11 00:23:35.871 INFO Successfully published block            slot: 5770916, graffiti: Some("Block by Stakely Lido07"), attestations: 99, deposits: 0, block_type: Blinded, service: block
```

Validator setup and version: Lighthouse/v3.4.0-38514c0
```
lighthouse vc
--beacon-nodes http://LH,http://SERVER_ON_MAINTENANCE,http://ip3
--network mainnet
--metrics
--metrics-allow-origin *
--graffiti "Block by xxxxxxx"
--suggested-fee-recipient 0x388Cxxxxxxxxxxxx
--init-slashing-protection
--metrics-port xxx
--http
--builder-proposals
```

Beacon setup and version: Lighthouse/v3.4.0-38514c0
```

      lighthouse beacon
      --network mainnet
      --execution-jwt  /xxxxxxxxxxxxxxx
      --execution-endpoint  http://localhost:xxxxx
      --subscribe-all-subnets
      --import-all-attestations
      --graffiti "Block by Stakely.io - blh01"
      --metrics
      --metrics-allow-origin xxxxx
      --http
      --http-address xxxxxxxx
      --http-port xxxxx
      --port xxxxxxx
      --checkpoint-sync-url https://mainnet-checkpoint-sync.stakely.io/
      --suggested-fee-recipient  0xxxxxxxxxx
      --builder http://xxxxxxx
```

Execution setup and version: Geth/v1.10.26-stable-e5eb32ac/linux-amd64/go1.18.8
```

      --mainnet
      --syncmode=snap
      --http
      --http.api "eth,net,web3,engine,admin"
      --http.corsdomain "*"
      --http.addr 127.0.0.1
      --http.port xxxx
      --cache  16384
      --port xxxx
      --authrpc.port xxx
      --authrpc.jwtsecret "xxxxxx"
      --authrpc.vhosts "xxxx"
```
Lighhouse Beacon logs are relevant  because the first lines show the BN loses several seconds because a new  block received changes the parent root of the Prepared beacon proposer for slot 5770016.

[lh-bn-orfandblock10022023.txt](https://github.com/sigp/lighthouse/files/10713019/lh-bn-orfandblock10022023.txt)

[mev-boost-orfandblock10022023.txt](https://github.com/sigp/lighthouse/files/10713027/mev-boost-orfandblock10022023.txt)

Important note: VC is connected to several BNs, one of the then is shut down because it is currently undergoing maintenance.This doesnt affect this problem but you'll see warnings in VC logs. 

[lh-vc-orfandblock10022023.txt](https://github.com/sigp/lighthouse/files/10713026/lh-vc-orfandblock10022023.txt)

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Closes #3963 (hopefully)

## Proposed Changes

Compute attestation selection proofs gradually each slot rather than in a single `join_all` at the start of each epoch. On a machine with 5k validators this replaces 5k tasks signing 5k proofs with 1 task that signs 5k/32 ~= 160 proofs each slot.

Based on testing with Goerli validators this seems to reduce the average time to produce a signature by preventing Tokio and the OS from falling over each other trying to run hundreds of threads. My testing so far has been with local keystores, which run on a dynamic pool of up to 512 OS threads because they use [`spawn_blocking`](https://docs.rs/tokio/1.11.0/tokio/task/fn.spawn_blocking.html) (and we haven't changed the default).

An earlier version of this PR hyper-optimised the time-per-signature metric to the detriment of the entire system's performance (see the reverted commits). The current PR is conservative in that it avoids touching the attestation service at all. I think there's more optimising to do here, but we can come back for that in a future PR rather than expanding the scope of this one.

The new algorithm for attestation selection proofs is:

- We sign a small batch of selection proofs each slot, for slots up to 8 slots in the future. On average we'll sign one slot's worth of proofs per slot, with an 8 slot lookahead.
- The batch is signed halfway through the slot when there is unlikely to be contention for signature production (blocks are <4s, attestations are ~4-6 seconds, aggregates are 8s+).

## Performance Data

_See first comment for updated graphs_.

Graph of median signing times before this PR:

signing_times_median

Graph of update attesters metric (includes selection proof signing) before this PR:

update_attesters_store

Median signing time after this PR (prototype from 12:00, updated version from 13:30):

signing_times_median_updated

99th percentile on signing times (bounded attestation signing from 16:55, now removed):

signing_times_99pc

Attester map update timing after this PR:

update_attesters_store_updated

Selection proof signings per second change:

signing_attempts

## Link to late blocks

I believe this is related to the slow block signings because logs from Stakely in #3963 show these two logs almost 5 seconds apart:

> Feb 23 18:56:23.978 INFO Received unsigned block, slot: 5862880, service: block, module: validator_client::block_service:393
> Feb 23 18:56:28.552 INFO Publishing signed block, slot: 5862880, service: block, module: validator_client::block_service:416

The only thing that happens between those two logs is the signing of the block:

https://github.com/sigp/lighthouse/blob/0fb58a680d6f0c9f0dc8beecf142186debff9a8d/validator_client/src/block_service.rs#L410-L414

Helpfully, Stakely noticed this issue without any Lighthouse BNs in the mix, which pointed to a clear issue in the VC.

## TODO

- [x] Further testing on testnet infrastructure.
- [x] Make the attestation signing parallelism configurable.

### CODE REVIEW

**Comment by michaelsproul** [comment]
## Updated perf data

This heatmap shows the minimum delay for attestations made by 45K validators across multiple VCs, as seen by a single BN. The first changes were introduced just after 03/01 15:00 (pardon the Americanism). We can see that the delay spikes terribly because of the overly restricted attestation signing parallelism. After reverting those changes and re-deploying to all nodes we see a return to normality, and a slight improvement! Although the 1-2.5s bucket is still the most populous there are also more attestations in the buckets below, particularly the 500-1ms bucket. There are also slightly less in the higher buckets. My explanation for this is that the selection proof signing has been moved "out of the way" and no longer interferes with attestations at the start of the epoch.

first_unaggregated_attestation

----

Here's another view of the same delay, this time showing the median. We can see the botched optimisation spiking the value, and then the fixed version improving the median delay, from ~1.65s to ~1.35s.

first_unaggregated_attestation_median

----

For completeness here's the 95th percentile of the delay. The earlier spike is due to the botched optimisation running on a single VC. Here we don't see as much improvement because some attestations still clogged behind each other (future work).

first_unaggregated_attestation_95pc

----

Finally here are the signing times that I was obsessed with hyper-optimising initially. We can see how the botched optimisation drops them (at the same time it killed attestation throughput) and the improved optimisation decreases them substantially.

signing_times_99pc_updated

And the median:

signing_times_median_updated

File: validator_client/src/block_service.rs [hunk 1]
```diff
@@ -407,17 +407,22 @@ impl<T: SlotClock + 'static, E: EthSpec> BlockService<T, E> {
...
             .await
             .map_err(|e| BlockError::Recoverable(format!("Unable to sign block: {:?}", e)))?;
+        let signing_time_ms =
+            Duration::from_secs_f64(signing_timer.map_or(0.0, |t| t.stop_and_record())).as_millis();
 
         info!(
             log,
             "Publishing signed block";
             "slot" => slot.as_u64(),
+            "signing_time_ms" => signing_time_ms,
```
Comment (paulhauner) [inline validator_client/src/block_service.rs:423]: Nice addition!

File: validator_client/src/duties_service.rs [hunk 2]
```diff
@@ -785,9 +787,144 @@ async fn poll_beacon_attesters_for_epoch<T: SlotClock + 'static, E: EthSpec>(
...
+        if let Some(duration) = slot_clock.duration_to_next_slot() {
+            sleep(duration.saturating_sub(slot_offset)).await;
+
+            let Some(current_slot) = slot_clock.now() else {
+                continue;
+            };
+
+            let lookahead_slot = current_slot + SELECTION_PROOF_SLOT_LOOKAHEAD;
+
+            let mut relevant_duties = duties_by_slot.split_off(&lookahead_slot);
```
Comment (paulhauner) [inline validator_client/src/duties_service.rs:835]: The `split_off` method on `BTreeMap` is really nice here.

**Review by paulhauner — APPROVED** [review]
Looks good, no changes just compliments!

It's great to see this getting implemented, it's a much better way to do things IMO.

**Comment by michaelsproul** [comment]
Thanks Paul! :blush: 

bors r+

**Pull request closed without merging**